File: db_connection.py
"""
Supabase Database Connection Module
Provides connection pooling and query execution for the Creative Acceleration Lab
"""

import os
from typing import Optional, Dict, List, Any
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

class DatabaseConnection:
    """
    Manages database connections to Supabase using the session pooler
    """

    def __init__(self):
        self.database_url = os.getenv('DATABASE_URL')
        self.supabase_url = os.getenv('SUPABASE_URL')
        self._connection = None

        if not self.database_url:
            raise ValueError("DATABASE_URL environment variable not set. Check your .env file.")

    def get_connection_string(self) -> str:
        """Returns the pooled connection string"""
        return self.database_url

    def connect_with_psycopg2(self):
        """
        Create a connection using psycopg2

        Usage:
            db = DatabaseConnection()
            conn = db.connect_with_psycopg2()
            cursor = conn.cursor()
            cursor.execute("SELECT * FROM your_table")
            results = cursor.fetchall()
            cursor.close()
            conn.close()
        """
        try:
            import psycopg2
            conn = psycopg2.connect(self.database_url)
            return conn
        except ImportError:
            raise ImportError("psycopg2 not installed. Run: pip install psycopg2-binary")
        except Exception as e:
            raise ConnectionError(f"Failed to connect to database: {e}")

    def connect_with_sqlalchemy(self):
        """
        Create a SQLAlchemy engine

        Usage:
            db = DatabaseConnection()
            engine = db.connect_with_sqlalchemy()

            # With pandas
            import pandas as pd
            df = pd.read_sql("SELECT * FROM your_table", engine)

            # With SQLAlchemy ORM
            from sqlalchemy.orm import sessionmaker
            Session = sessionmaker(bind=engine)
            session = Session()
        """
        try:
            from sqlalchemy import create_engine
            engine = create_engine(self.database_url)
            return engine
        except ImportError:
            raise ImportError("sqlalchemy not installed. Run: pip install sqlalchemy")
        except Exception as e:
            raise ConnectionError(f"Failed to create SQLAlchemy engine: {e}")

    async def connect_with_asyncpg(self):
        """
        Create an async connection pool using asyncpg

        Usage:
            import asyncio

            async def main():
                db = DatabaseConnection()
                pool = await db.connect_with_asyncpg()

                async with pool.acquire() as conn:
                    result = await conn.fetch("SELECT * FROM your_table")
                    print(result)

                await pool.close()

            asyncio.run(main())
        """
        try:
            import asyncpg
            pool = await asyncpg.create_pool(self.database_url)
            return pool
        except ImportError:
            raise ImportError("asyncpg not installed. Run: pip install asyncpg")
        except Exception as e:
            raise ConnectionError(f"Failed to create async connection pool: {e}")

    def test_connection(self) -> bool:
        """
        Test if the database connection works

        Returns:
            bool: True if connection successful, False otherwise
        """
        try:
            conn = self.connect_with_psycopg2()
            cursor = conn.cursor()
            cursor.execute("SELECT 1")
            result = cursor.fetchone()
            cursor.close()
            conn.close()
            return result[0] == 1
        except Exception as e:
            print(f"Connection test failed: {e}")
            return False


# Convenience function for quick connections
def get_db_connection():
    """
    Quick helper to get a database connection

    Usage:
        from db_connection import get_db_connection

        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute("SELECT * FROM your_table")
        results = cursor.fetchall()
        cursor.close()
        conn.close()
    """
    db = DatabaseConnection()
    return db.connect_with_psycopg2()


def get_db_engine():
    """
    Quick helper to get a SQLAlchemy engine

    Usage:
        from db_connection import get_db_engine
        import pandas as pd

        engine = get_db_engine()
        df = pd.read_sql("SELECT * FROM your_table", engine)
    """
    db = DatabaseConnection()
    return db.connect_with_sqlalchemy()


if __name__ == "__main__":
    # Test the connection
    print("Testing database connection...")
    db = DatabaseConnection()

    if db.test_connection():
        print("✅ Database connection successful!")
        print(f"Connected to: {db.supabase_url}")
    else:
        print("❌ Database connection failed!")
        print("Please check your .env file and credentials.")

File: db_connection.js
/**
 * Supabase Database Connection Module (Node.js)
 * Provides connection pooling and query execution for the Creative Acceleration Lab
 */

require('dotenv').config();
const { Pool } = require('pg');

class DatabaseConnection {
    constructor() {
        this.connectionString = process.env.DATABASE_URL;
        this.supabaseUrl = process.env.SUPABASE_URL;
        this.pool = null;

        if (!this.connectionString) {
            throw new Error('DATABASE_URL environment variable not set. Check your .env file.');
        }
    }

    /**
     * Get a connection pool
     * Reuses the same pool for efficiency
     */
    getPool() {
        if (!this.pool) {
            this.pool = new Pool({
                connectionString: this.connectionString,
                ssl: {
                    rejectUnauthorized: false
                },
                max: parseInt(process.env.MAX_CONNECTIONS || '20'),
                idleTimeoutMillis: 30000,
                connectionTimeoutMillis: 2000,
            });

            // Handle pool errors
            this.pool.on('error', (err) => {
                console.error('Unexpected error on idle client', err);
            });
        }

        return this.pool;
    }

    /**
     * Execute a query
     *
     * @param {string} queryText - SQL query
     * @param {Array} params - Query parameters
     * @returns {Promise<Object>} Query result
     *
     * Usage:
     *   const db = new DatabaseConnection();
     *   const result = await db.query('SELECT * FROM users WHERE id = $1', [userId]);
     *   console.log(result.rows);
     */
    async query(queryText, params = []) {
        const pool = this.getPool();
        try {
            const result = await pool.query(queryText, params);
            return result;
        } catch (error) {
            console.error('Query error:', error);
            throw error;
        }
    }

    /**
     * Execute a transaction
     *
     * @param {Function} callback - Async function that receives a client
     * @returns {Promise<any>} Transaction result
     *
     * Usage:
     *   const db = new DatabaseConnection();
     *   await db.transaction(async (client) => {
     *       await client.query('INSERT INTO users (name) VALUES ($1)', ['John']);
     *       await client.query('INSERT INTO logs (action) VALUES ($1)', ['User created']);
     *   });
     */
    async transaction(callback) {
        const pool = this.getPool();
        const client = await pool.connect();

        try {
            await client.query('BEGIN');
            const result = await callback(client);
            await client.query('COMMIT');
            return result;
        } catch (error) {
            await client.query('ROLLBACK');
            console.error('Transaction error:', error);
            throw error;
        } finally {
            client.release();
        }
    }

    /**
     * Test the database connection
     *
     * @returns {Promise<boolean>} True if connection successful
     */
    async testConnection() {
        try {
            const result = await this.query('SELECT 1 as test');
            return result.rows[0].test === 1;
        } catch (error) {
            console.error('Connection test failed:', error);
            return false;
        }
    }

    /**
     * Close the connection pool
     * Call this when shutting down the application
     */
    async close() {
        if (this.pool) {
            await this.pool.end();
            this.pool = null;
        }
    }
}

// Singleton instance for convenience
let dbInstance = null;

/**
 * Get the database singleton instance
 *
 * Usage:
 *   const { getDB } = require('./db_connection');
 *   const db = getDB();
 *   const result = await db.query('SELECT * FROM users');
 */
function getDB() {
    if (!dbInstance) {
        dbInstance = new DatabaseConnection();
    }
    return dbInstance;
}

// Test connection if run directly
if (require.main === module) {
    (async () => {
        console.log('Testing database connection...');
        const db = new DatabaseConnection();

        const isConnected = await db.testConnection();

        if (isConnected) {
            console.log('✅ Database connection successful!');
            console.log(`Connected to: ${db.supabaseUrl}`);
        } else {
            console.log('❌ Database connection failed!');
            console.log('Please check your .env file and credentials.');
        }

        await db.close();
        process.exit(isConnected ? 0 : 1);
    })();
}

module.exports = {
    DatabaseConnection,
    getDB
};
